File: src/main/java/fr/raconteur/simpleskinswapper/networking/SkinRefreshPayload.java
package fr.raconteur.simpleskinswapper.networking;

import com.mojang.authlib.properties.Property;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;

public record SkinRefreshPayload(Property textureProperty) implements CustomPacketPayload {
    public static final CustomPacketPayload.Type<SkinRefreshPayload> PACKET_ID =
            new CustomPacketPayload.Type<>(Identifier.fromNamespaceAndPath("skinshuffle", "skin_refresh"));

    public static final StreamCodec<RegistryFriendlyByteBuf, SkinRefreshPayload> PACKET_CODEC =
            StreamCodec.ofMember(SkinRefreshPayload::write, SkinRefreshPayload::read);

    private static void write(SkinRefreshPayload payload, RegistryFriendlyByteBuf buf) {
        Property p = payload.textureProperty();
        buf.writeBoolean(p.hasSignature());
        buf.writeUtf(p.name());
        buf.writeUtf(p.value());
        if (p.hasSignature()) buf.writeUtf(p.signature());
    }

    private static SkinRefreshPayload read(RegistryFriendlyByteBuf buf) {
        boolean hasSig = buf.readBoolean();
        String name = buf.readUtf();
        String value = buf.readUtf();
        String sig = hasSig ? buf.readUtf() : null;
        return new SkinRefreshPayload(new Property(name, value, sig));
    }

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/networking/SkinRefreshPayload.kt
package fr.raconteur.simpleskinswapper.networking

import com.mojang.authlib.properties.Property
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier

data class SkinRefreshPayload(val textureProperty: Property) : CustomPacketPayload {

    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = PACKET_ID

    private fun write(buf: RegistryFriendlyByteBuf) {
        buf.writeBoolean(textureProperty.hasSignature())
        buf.writeUtf(textureProperty.name())
        buf.writeUtf(textureProperty.value())
        if (textureProperty.hasSignature()) buf.writeUtf(textureProperty.signature()!!)
    }

    companion object {
        @JvmField
        val PACKET_ID: CustomPacketPayload.Type<SkinRefreshPayload> =
            CustomPacketPayload.Type(Identifier.fromNamespaceAndPath("skinshuffle", "skin_refresh"))

        @JvmField
        val PACKET_CODEC: StreamCodec<RegistryFriendlyByteBuf, SkinRefreshPayload> =
            StreamCodec.ofMember(SkinRefreshPayload::write, Companion::read)

        private fun read(buf: RegistryFriendlyByteBuf): SkinRefreshPayload {
            val hasSig = buf.readBoolean()
            val name = buf.readUtf()
            val value = buf.readUtf()
            val sig = if (hasSig) buf.readUtf() else null
            return SkinRefreshPayload(Property(name, value, sig))
        }
    }
}
